### PR TITLE
Button Group

### PR DESCRIPTION
**Pull request by dannify**

### CODE REVIEW

File: packages/@adobe/spectrum-css-temp/components/dialog/index.css [hunk 1]
```diff
@@ -192,10 +192,6 @@ governing permissions and limitations under the License.
   justify-content: flex-end;
   /* this padding should be safe as button group is always end aligned */
   padding-inline-start: var(--spectrum-dialog-gap-size);
-
-  &.spectrum-Dialog-buttonGroup--noFooter {
```
Comment (dannify) [inline packages/@adobe/spectrum-css-temp/components/dialog/index.css:196]: This messes up the width of the buttons when vertical. @snowystinger please halp

File: packages/@adobe/spectrum-css-temp/components/buttongroup/index.css [hunk 2]
```diff
@@ -3,7 +3,6 @@ Copyright 2019 Adobe. All rights reserved.
 This file is licensed to you under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License. You may obtain a copy
 of the License at http://www.apache.org/licenses/LICENSE-2.0
-
```
Comment (snowystinger) [inline packages/@adobe/spectrum-css-temp/components/buttongroup/index.css:6]: I think this is supposed to stay?

File: packages/@adobe/spectrum-css-temp/components/buttongroup/index.css [hunk 3]
```diff
@@ -15,23 +14,32 @@ governing permissions and limitations under the License.
 }
 
 .spectrum-ButtonGroup {
-  display: flex;
+  display: inline-flex;
+  /* Added this so it would wrap */
+  flex-wrap: wrap;
+  align-items: flex-end;
```
Comment (snowystinger) [inline packages/@adobe/spectrum-css-temp/components/buttongroup/index.css:20]: should all button groups flex end? or should this be something from slots?
maybe flex end is a good default and if needed, people can do flex-start through a new class

File: packages/@adobe/spectrum-css-temp/components/buttongroup/index.css [hunk 4]
```diff
@@ -15,23 +14,32 @@ governing permissions and limitations under the License.
...
 
   .spectrum-Button,
-  .spectrum-ActionButton {
+  .spectrum-ActionButton,
+  .spectrum-Tool {
     flex-shrink: 0;
   }
 
+  > .spectrum-Rule--vertical {
```
Comment (snowystinger) [inline packages/@adobe/spectrum-css-temp/components/buttongroup/index.css:28]: spectrum rule shouldn't be in this file? nor should tool, action button, or button. They'll have different names than their counterparts in the the button css file and the divider css file

File: packages/@adobe/spectrum-css-temp/components/buttongroup/index.css [hunk 5]
```diff
@@ -15,23 +14,32 @@ governing permissions and limitations under the License.
...
+    block-size: auto;
```
Comment (snowystinger) [inline packages/@adobe/spectrum-css-temp/components/buttongroup/index.css:29]: why is this here?

File: packages/@adobe/spectrum-css-temp/components/buttongroup/index.css [hunk 6]
```diff
@@ -15,23 +14,32 @@ governing permissions and limitations under the License.
...
+    align-self: stretch;
+  }
+
   .spectrum-Button + .spectrum-Button {
-    margin-left: var(--spectrum-buttongroup-button-gap-x);
+    margin-inline-start: var(--spectrum-buttongroup-button-gap-x);
```
Comment (snowystinger) [inline packages/@adobe/spectrum-css-temp/components/buttongroup/index.css:34]: @devongovett can we use your Flex gap stuff?

File: packages/@react-spectrum/button/src/Button.tsx [hunk 7]
```diff
@@ -27,7 +27,7 @@ let VARIANT_MAPPING = {
 
 function Button(props: SpectrumButtonProps, ref: FocusableRef) {
   props = useProviderProps(props);
-  props = useSlotProps(props);
+  props = useSlotProps(props, 'button');
```
Comment (snowystinger) [inline packages/@react-spectrum/button/src/Button.tsx:30]: is this going to conflict with Devon's PR?

File: packages/@react-spectrum/buttongroup/src/ButtonGroup.tsx [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+      if (scale !== lastScale) {
+        if (lastScale === 'large' && scale === 'medium') {
+          setHasOverflow(false);
+        } else {
+          checkForOverflow();
+        }
+      }
+      lastScale = scale;
+    }
+  }, [setHasOverflow, checkForOverflow, scale, orientation]);
```
Comment (snowystinger) [inline packages/@react-spectrum/buttongroup/src/ButtonGroup.tsx:61]: I think this will have some trouble if children are added/removed or if the locale changes or if any strings change

example where that might happen, confirm button in dialog is pressed, it performs an async action before closing, and while it does that, it displays a spinner and says wait while disabled. If the buttons were horizontal and the wait made it wrap, then we'd end up in a situation that doesn't look very good

File: packages/@react-spectrum/buttongroup/src/ButtonGroup.tsx [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+  useEffect(() => {
+    if (orientation !== 'vertical') {
+      // I think performance could be optimized here by creating a global, debounced hook for listening to resize
```
Comment (snowystinger) [inline packages/@react-spectrum/buttongroup/src/ButtonGroup.tsx:65]: we should probably make a useDebouncedResize hook or something with a single listener in it, we can probably put that off for now though, right now it's buttongroup, tabs, breadcrumbs, none of which are likely to have a lot of instances on a page at once

File: packages/@react-spectrum/buttongroup/src/ButtonGroup.tsx [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+            'spectrum-ButtonGroup--vertical': hasOverflow || orientation === 'vertical'
+          },
+          styleProps.className
+        )
+      }>
+      <SlotProvider
+        slots={{
+          button: {
+            isDisabled,
+            UNSAFE_className: classNames(styles, 'spectrum-Button')
```
Comment (snowystinger) [inline packages/@react-spectrum/buttongroup/src/ButtonGroup.tsx:94]: this classname should be renamed to `spectrum-ButtonGroup-button`

File: packages/@react-spectrum/buttongroup/test/ButtonGroup.test.js [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+let buttonGroupId = 'button-group';
+let onPressSpy1 = jest.fn();
+let onPressSpy2 = jest.fn();
+let onPressSpy3 = jest.fn();
+
+function renderComponent(Component, props = {}) {
+  if (Component === V2ButtonGroup) {
+    return render(
+      <Provider theme={theme}>
+        <V2ButtonGroup data-testid={buttonGroupId} {...props}>
```
Comment (snowystinger) [inline packages/@react-spectrum/buttongroup/test/ButtonGroup.test.js:38]: are any tests testing this? i think we can get rid of the V2 parity, it's pretty different in implementation no?

File: packages/@react-types/buttongroup/src/index.d.ts [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+import {ReactNode} from 'react';
+
+export interface SpectrumButtonGroupProps extends DOMProps, StyleProps {
+  /** Whether the ButtonGroup is disabled. */
+  isDisabled?: boolean,
+  /**
+   * The axis the ButtonGroup should align with.
+   * @default "horizontal"
+   */
+  orientation?: Orientation,
```
Comment (snowystinger) [inline packages/@react-types/buttongroup/src/index.d.ts:24]: Should it be mentioned that if vertical is used, no reflow will happen, it'll just always be vertical. But if horizontal is used, then reflow to vertical will happen?

File: packages/@react-spectrum/buttongroup/src/ButtonGroup.tsx (hunk 8 above)
Comment (snowystinger) [reply packages/@react-spectrum/buttongroup/src/ButtonGroup.tsx:61]: We may need to use a MutationObserver https://caniuse.com/#feat=mutationobserver because the internal change in the confirm button to display a wait spinner may not trigger anything here

Comment (snowystinger) [reply packages/@react-spectrum/buttongroup/src/ButtonGroup.tsx:61]: When a change occurs, we'll need to remove the vertical class, useLayoutEffect to see if there's overflow and quickly reapply the vertical class if there is overflow

File: packages/@adobe/spectrum-css-temp/components/dialog/index.css [hunk 13]
```diff
@@ -198,6 +198,8 @@ governing permissions and limitations under the License.
   }
 }
 
+
+
```
Comment (devongovett) [inline packages/@adobe/spectrum-css-temp/components/dialog/index.css:202]: remove?

File: packages/@adobe/spectrum-css-temp/components/buttongroup/index.css (hunk 4 above)
Comment (LFDanLu) [reply packages/@adobe/spectrum-css-temp/components/buttongroup/index.css:28]: kk removed

File: packages/@react-spectrum/buttongroup/test/ButtonGroup.test.js (hunk 11 above)
Comment (LFDanLu) [reply packages/@react-spectrum/buttongroup/test/ButtonGroup.test.js:38]: I can remove the v2 stuff

File: packages/@react-types/buttongroup/src/index.d.ts (hunk 12 above)
Comment (LFDanLu) [reply packages/@react-types/buttongroup/src/index.d.ts:24]: I think this should be documented in the docs, I'll say something like "See the docs for reflow behavior".

Comment (LFDanLu) [reply packages/@react-types/buttongroup/src/index.d.ts:24]: will be easier to link to it from here

File: packages/@react-spectrum/buttongroup/README.md [hunk 14]
```diff
@@ -0,0 +1,3 @@
+# @react-spectrum/button
```
Comment (devongovett) [inline packages/@react-spectrum/buttongroup/README.md:1]:
```suggestion
# @react-spectrum/buttongroup
```
